test/test_transformers.py

```diff
@@ -2495,6 +2495,53 @@ def test_cudnn_attention_trivial_output_transpose(self, device):
         o.backward(o)
         torch.testing.assert_close(x.grad, x_cpu.grad.cuda(), atol=7e-3, rtol=7e-3)
 
+    @skipIfRocm
+    @unittest.skipIf(not PLATFORM_SUPPORTS_CUDNN_ATTENTION, "cudnn Attention is not supported on this system")
+    def test_cudnn_attention_preserves_query_layout(self, device):
+        class Config:
+            n_embd: int = 512
+            n_head: int = 8
+            n_layer: int = 6
+            n_ctx: int = 2048
+            bias: bool = False
+
+        class CausalSelfAttention(nn.Module):
+
+            def __init__(self, config):
+                super().__init__()
+                assert config.n_embd % config.n_head == 0
+                # key, query, value projections for all heads, but in a batch
+                self.c_attn = nn.Linear(config.n_embd, 3 * config.n_embd, bias=config.bias)
+                # output projection
+                self.c_proj = nn.Linear(config.n_embd, config.n_embd, bias=config.bias)
+                self.n_head = config.n_head
+                self.n_embd = config.n_embd
+
+            def forward(self, x):
+                B, T, C = x.size() # batch size, sequence length, embedding dimensionality (n_embd)
+                q, k, v  = self.c_attn(x).split(self.n_embd, dim=2)
+                k = k.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
+                q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
+                v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
+
+                y = F.scaled_dot_product_attention(q, k, v, attn_mask=None, is_causal=True)
+
+                # HERE, WE NEED THIS CONTIGUOUS TO BE A NO-OP
+                # y = y.transpose(1, 2).contiguous().view(B, T, C)
+                y = y.transpose(1, 2).view(B, T, C)
```

Comment:
strictly speaking, this doesn't check that `.contiguous` is a no-op, it just checks that dimension subspaces are correct (e.g. B could still be sparse). I'd directly check expected strides.

Reply:
agreed, I think sweeping over a bunch of permutations would be helpful:
https://github.com/pytorch/pytorch/blob/853da168fc17c689b652edb5998c986294b9ecbf/test/inductor/test_flex_attention.py#L1905